ARCHITECTURE.md

```diff
@@ -0,0 +1,196 @@
+- The libraries convert pagination APIs into C++ iterators.
+- The libraries can be configured to log RPCs and responses, to help application
+  developers troubleshoot their code.
+
+## Where is the code?
+
+Each library is found in a subdirectory of [/google/cloud/](/google/cloud). The
+name of the subdirectory is chosen to match the most distinctive part of the
+service name. For example, `/google/cloud/storage` contains the code for the
+[Google Cloud Storage](https://cloud.google.com/storage) service.
+
+Within each directory you will find (almost always) the same structure:
+
+- `${library}/*.h` are the public headers for the library. The same directory
+  contains the `${library}/*.cc` implementation files and `*_test.cc` for unit
+  tests. More often than not all the unit tests for `foo.{h,cc}` are found in
+  `foo_test.cc`, but there may be exceptions when the tests grow too large to
+  fit in a single file (we have lots of tests!).
+- `mocks/` also contains public APIs, typically this is a header-only library
+  defining mocks so application developers can mock the library in their own
+  tests. Note that some libraries do not have mocks.
+- `quickstart/` contains a small program showing how to use the library in a
+  larger project. This is intended for application developers wanting to, well,
+  use the libraries quickly.
+- `samples/` (sometimes `examples/`) contain code snippets that are extracted
+  both into the Doxygen reference documentation and to the cloud.google.com
+  site, as part of the service documentation. Some libraries have a single
+  program in this directory, others have smaller programs (or more samples).
+
+The remaining directories are implementation details, or at least only intended
+to be interesting for `google-cloud-cpp` developers and contributors.
+
+- `internal/` contains implementation details for the library. The files in
+  these directories are *not* intended for public consumption, but contributors
+  and developers need to change them, of course. The relationship between `.h`,
+  `.cc` and `_test.cc` files mirrors the one found in `${library}/`.
+- `benchmarks/` contains "macro benchmarks" for the library, typically programs
+  that interact with the production environment to verify the library meets the
+  latency, throughput, and CPU overhead (as in "low CPU overhead") requirements.
+- `ci/` may contain helper scripts used in the CI builds, for example, to
+  startup and shutdown any emulators used by the library.
+- `doc/` contains the landing page for any Doxygen documentation, as well as
+  additional Doxygen docs that are not extracted from the code comments.
+- `integration_tests/` (or simply `tests/` in some libraries) contain any
+  integration tests where we test the library using emulators and/or the
+  production environment.
+- `testing/` contains testing utilities shared by the unit tests and the
+  integration tests of the library. They are not intended to be part of the
```

Comment:
"part of the \<something>"?

Reply:
"public API", but this is implied in the heading, so I am removing it.